From 2ceeb48c90b547c766e67221c0bbd9c15237cfd0 Mon Sep 17 00:00:00 2001
From: Kilian von Pflugk <github@jumoog.io>
Date: Sat, 10 Aug 2024 14:33:05 +0200
Subject: [PATCH] add option for allowed destinations

---
 net/socks5/socks5.go | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/net/socks5/socks5.go b/net/socks5/socks5.go
index 0d651537f..640880572 100644
--- a/net/socks5/socks5.go
+++ b/net/socks5/socks5.go
@@ -21,9 +21,11 @@
 	"io"
 	"log"
 	"net"
+	"regexp"
 	"strconv"
 	"time"
 
+	"tailscale.com/envknob"
 	"tailscale.com/types/logger"
 )
 
@@ -183,6 +185,19 @@ func (c *Conn) handleRequest() error {
 		return err
 	}
 
+	// whitelist destinations
+	if s := envknob.String("TS_SOCKS_ALLOWED_DESTINATION"); s != "" {
+		match, _ := regexp.MatchString(s, req.destination.addr)
+		if !match {
+			res := errorResponse(hostUnreachable)
+			buf, _ := res.marshal()
+			c.clientConn.Write(buf)
+			return fmt.Errorf("%v is not allowed", req.destination)
+		} else {
+			log.Printf("allow access to: %v on port: %v", req.destination.addr, req.destination.port)
+		}
+	}
+
 	c.request = req
 	switch req.command {
 	case connect:
-- 
2.46.0.windows.1

