diff --git a/net/socks5/socks5.go b/net/socks5/socks5.go
index f67dc1e..ba83942 100644
--- a/net/socks5/socks5.go
+++ b/net/socks5/socks5.go
@@ -13,6 +13,7 @@ package socks5
 package socks5

 import (
+	"regexp"
 	"bytes"
 	"context"
 	"encoding/binary"
@@ -25,6 +26,7 @@ import (
 	"strconv"
 	"time"

+	"tailscale.com/envknob"
 	"tailscale.com/types/logger"
 )

@@ -191,6 +193,19 @@ func (c *Conn) handleRequest() error {
 		return err
 	}

+	// whitelist destinations
+	if s := envknob.String("TS_SOCKS_ALLOWED_DESTINATION"); s != "" {
+		match, _ := regexp.MatchString(s, req.destination.addr)
+		if !match {
+			res := errorResponse(hostUnreachable)
+			buf, _ := res.marshal()
+			c.clientConn.Write(buf)
+			return fmt.Errorf("%v is not allowed", req.destination)
+		} else {
+			log.Printf("allow access to: %v on port: %v", req.destination.addr, req.destination.port)
+		}
+	}
+
 	c.request = req
 	switch req.command {
 	case connect:
